### PR TITLE
Strong Naming

### PR DESCRIPTION
**Pull request by ScottGuymer**
Added signing keys but still need to update dependencies to be signed.

https://github.com/PrometheusClientNet/Prometheus.Client/issues/55

### CODE REVIEW

**Comment by ScottGuymer** [comment]
@phnx47 It looks like this repo hasnt been updated to reference the changes in Prometheus.Client 3.0.

Do you know if anyone else is working on this update?

**Comment by ScottGuymer** [comment]
Looks like there are a lot of changes in Prometheus.Client that break this package.

Firstly Prometheus.Client doesnt build a net45 target anymore. Which there are a few compiler directives that rely on. I am kina presuming that this is to allow it to do different thinks based on whether it is running full framework or core. 

Im not sure how  i achieve this with only netstandard options.

**Comment by ScottGuymer** [comment]
@phnx47 i have fixed up the issues with the changes in 3.0 (i think)

Would be grateful if you could look over my changes.

The tests are still failing due to the way they are written and they share a metrics registry across tests.
If you are happy with the approach i will try and refactor them more to fix them up

**Comment by ScottGuymer** [comment]
It also needs like the tests fail but dont fail the build.......

File: src/Prometheus.Client.MetricServer/BaseMetricServer.cs [hunk 1]
```diff
@@ -20,19 +21,21 @@ public abstract class BaseMetricServer
         /// <param name="registry">Collector registry</param>
         /// <param name="collectors">IOnDemandCollectors</param>
         /// <param name="useDefaultCollectors">Use default collectors</param>
-        protected BaseMetricServer(ICollectorRegistry registry, List<IOnDemandCollector> collectors, bool useDefaultCollectors)
+        protected BaseMetricServer(ICollectorRegistry registry, List<ICollector> collectors, bool useDefaultCollectors)
```
Comment (sanych-sun) [inline src/Prometheus.Client.MetricServer/BaseMetricServer.cs]: Collectors parameter could be removed, as client code can modify ICollectorRegistry directly (it was not possible in v2, so we had to have extra parameter)

File: src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿#if NET45
+#if NET47
```
Comment (sanych-sun) [inline src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs]: Most likely it should be netstandard2.0 directive instead, as HttpListerner is available in netstandard 2.0

File: src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs [hunk 3]
```diff
@@ -139,7 +139,7 @@ private void StartListen()
                         response.ContentType = Defaults.ContentType;
                         using (var outputStream = response.OutputStream)
                         {
-                            ScrapeHandler.Process(Registry, outputStream);
+                            ScrapeHandler.ProcessAsync(Registry, outputStream);
```
Comment (sanych-sun) [inline src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs]: It should await. Either with .GetAwaiter().GetResult() or as a async/await method.

File: src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj [hunk 4]
```diff
@@ -5,8 +5,8 @@
     <AssemblyTitle>Prometheus.Client.MetricServer</AssemblyTitle>
     <VersionPrefix>2.1.1</VersionPrefix>
     <Authors>Sergey Kuznetsov</Authors>
-    <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">netstandard1.3;netstandard2.0</TargetFrameworks>
-    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">net45;netstandard1.3;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">netstandard1.3;netstandard2.0;netcoreapp2.2;</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">net47;netstandard1.3;netstandard2.0;netcoreapp2.2;</TargetFrameworks>
```
Comment (sanych-sun) [inline src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj]: Do we need this target? As far as we have netstandard2.0

**Comment by sanych-sun** [comment]
Hi @ScottGuymer!

Thank you for the contributing. I've added some comments on your changes. Please let me know if you need any assistance to make it done, as there are lost of changes and we had no chance to update MetricServer as a part of v3 initiative :flushed:

File: src/Prometheus.Client.MetricServer/BaseMetricServer.cs (hunk 1 above)
Comment (ScottGuymer) [reply src/Prometheus.Client.MetricServer/BaseMetricServer.cs]: fixed!

File: src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs (hunk 2 above)
Comment (ScottGuymer) [reply src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs]: changed to NETSTANDARD20

File: src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs (hunk 3 above)
Comment (ScottGuymer) [reply src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs]: Done

File: src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj (hunk 4 above)
Comment (ScottGuymer) [reply src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj]: Just coming back to .net after a while away and im not sure!

I will take your lead on this and remove it!

**Comment by ScottGuymer** [comment]
@sanych-sun I have made changes based on your comments.

I am still struggling with the failing tests based on the refacotring of the way the Metrics registry works.

Could do with a pointer on how you might fix them? 

As far as i can see it it is just the way the tests are exectuted that is the issue and not the actual test cases failing. They are failing because it is re-using the defult registry and trying to re-add the default metrics which is resulting in an error because they already exist.

**Comment by ScottGuymer** [comment]
Im also not sure why the build isnt failing due to the broken tests cant see anything wrong in the appveyor config

File: src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs [hunk 5]
```diff
@@ -79,12 +91,12 @@ public MetricServer(string host, int port, string url, ICollectorRegistry regist
         /// <param name="port">Port</param>
         /// <param name="mapPath">Map Path: should start with '/'</param>
         /// <param name="registry">Collector registry</param>
-        /// <param name="collectors">IOnDemandCollectors</param>
+        /// <param name="collectors">ICollectors</param>
         /// <param name="useHttps">use Https</param>
         /// <param name="useDefaultCollectors">Use default collectors</param>
-        public MetricServer(string host, int port, string mapPath, ICollectorRegistry registry, List<IOnDemandCollector> collectors, bool useHttps,
+        public MetricServer(string host, int port, string mapPath, ICollectorRegistry registry, List<ICollector> collectors, bool useHttps,
```
Comment (sanych-sun) [inline src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs]: This also has to be removed

**Comment by sanych-sun** [comment]
@ScottGuymer in core library tests we do not use static API anymore. We replaced Metrics class usage, by explicit creation of registry and factory. Most likely you have to create a new metric server for each tests by passing new CollectorRegistry explicitly, otherwise it try to use static one.

**Comment by ScottGuymer** [comment]
I have adapted the MetricsServer with an optional param for the CollectorRegistry.

Not sure if this is the correct way to go with this. What do you think?

Im also running into issues with the tests failing because the MetricsServer isnt started by the time the tests try to do their acting. Cant see a simple way to rewrite it to make it awaitable..

File: src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs (hunk 5 above)
Comment (ScottGuymer) [reply src/Prometheus.Client.MetricServer/MetricServer.HttpListener.cs]: done

**Comment by ScottGuymer** [comment]
If i remove the HttpListener version of the MetricsServer and just use the kestrel one everything seems to work.

Will this work well for people using the full framework via IIS for istance?

**Comment by ScottGuymer** [comment]
Looking at this for netstandard 2.0 we can just use kestrel and give people the choice of running in process or out of process

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/?view=aspnetcore-2.2&tabs=windows

**Comment by ScottGuymer** [comment]
I was pondering if we should re-design the MetricsServer anyway.. As it is getting to be a bit of a beast with so many constructor overloads.

**Comment by sanych-sun** [comment]
> I was pondering if we should re-design the MetricsServer anyway.. As it is getting to be a bit of a beast with so many constructor overloads.

Nice idea, I'll prototype it tonight by introducing options object instead of parameters. It could help.

**Comment by sanych-sun** [comment]
I've created separate pull request to replace all constructors with single options class. Let's wait for the merge in master and rebase this PR.

BTW: to solve the problem with unit tests I've change code to start listener on main thread.

**Comment by ScottGuymer** [comment]
OK lets get that merged in so i can take care of the rebase which is going to be painful...

File: src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj [hunk 6]
```diff
@@ -1,11 +1,13 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <Description>MetricServer for the Prometheus.Client</Description>
     <Copyright>2019 © Sergey Kuznetsov</Copyright>
     <AssemblyTitle>Prometheus.Client.MetricServer</AssemblyTitle>
     <VersionPrefix>3.0.0</VersionPrefix>
     <Authors>Sergey Kuznetsov</Authors>
     <TargetFrameworks>netstandard1.3;netstandard2.0;netcoreapp2.2</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">netstandard1.3;netstandard2.0;netcoreapp2.2;</TargetFrameworks>
```
Comment (sanych-sun) [inline src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj]: This is unneeded anymore.

File: tests/Prometheus.Client.MetricServer.NetCore.Tests/MetricServerTests.cs [hunk 7]
```diff
@@ -87,13 +87,16 @@ public async Task MapPath(string mapPath)
         [Fact]
         public async Task FindMetric()
         {
+            var registry = new CollectorRegistry();
```
Comment (sanych-sun) [inline tests/Prometheus.Client.MetricServer.NetCore.Tests/MetricServerTests.cs]: Duplicated code, can be removed

File: tests/Prometheus.Client.MetricServer.NetCore.Tests/MetricServerTests.cs [hunk 8]
```diff
@@ -87,13 +87,16 @@ public async Task MapPath(string mapPath)
...
             const int port = 9000;
             var registry = new CollectorRegistry();
             var factory = new MetricFactory(registry);
             var metricServer = new MetricServer(registry, new MetricServerOptions { Port = port });
 
             metricServer.Start();
 
+            var metrics = Metrics.WithCustomRegistry(registry);
```
Comment (sanych-sun) [inline tests/Prometheus.Client.MetricServer.NetCore.Tests/MetricServerTests.cs]: Can be removed

File: src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj (hunk 6 above)
Comment (sanych-sun) [reply src/Prometheus.Client.MetricServer/Prometheus.Client.MetricServer.csproj]: Could you please revert all changes into csproj file except of SignAssembly stuff?

**Comment by sanych-sun** [comment]
@ScottGuymer almost there!!! Please remove some unneeded code (rebase artifacts) and PR finally could be merged.

**Comment by ScottGuymer** [comment]
Sorry i hadnt finished, just wanted to see the diff and got interrupted with a meeting 😆

**Comment by ScottGuymer** [comment]
Done and ready for review

**Comment by sanych-sun** [comment]
I've merged your changes. Thank you for the contribution and sorry for some mess in the code, we are about to release v3 on the library, which has a lot of changes especially for allocation optimization. We still have some more repositories to be updated so it will take some time. Also your feedback on v3 changes would be much appreciated.

Thanks!

**Comment by ScottGuymer** [comment]
Great thanks.

I will try and get around the other repos and add strong naming.

Can you publish this to nuget?

**Comment by phnx47** [comment]
@ScottGuymer Yes. Sorry for waiting. I will publish it today/tomorrow

**Comment by phnx47** [comment]
3.0.0-rc1

**Pull request merged**